File: Plugins/Orimath.Basics.View/ViewModels/EffectViewModel.cs
using System.IO;
using System.Windows.Input;
using Orimath.Controls;
using Orimath.Plugins;

namespace Orimath.Basics.View.ViewModels
{
    public class EffectViewModel : NotifyPropertyChanged
    {
        private readonly IEffect _effect;

        public ICommand EffectCommand { get; }

        public string Name => _effect.Name;

        public string ShortcutKey => _effect.ShortcutKey;

        public string ToolTip => string.IsNullOrWhiteSpace(ShortcutKey)
            ? Name
            : $"{Name} ({ShortcutKey})";

        public Stream IconStream => _effect.Icon;

        public EffectViewModel(IEffect effect, IMessenger messenger)
        {
            _effect = effect;
            EffectCommand = messenger.GetEffectCommand(effect);
        }
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/EffectViewModel.fs
namespace Orimath.Basics.View.ViewModels
open System
open Orimath.Controls
open Orimath.Plugins

type EffectViewModel(effect: IEffect, messenger: IMessenger) =
    inherit NotifyPropertyChanged()
    member val EffectCommand = messenger.GetEffectCommand(effect)
    member _.Name = effect.Name
    member _.ShortcutKey = effect.ShortcutKey
    member _.IconStream = effect.Icon
    member _.ToolTip =
        if String.IsNullOrWhiteSpace(effect.ShortcutKey)
        then effect.Name
        else $"%s{effect.Name} (%s{effect.ShortcutKey})"


type EffectListViewModel(workspace: IWorkspace, messenger: IMessenger) =
    let lazyEffect =
        lazy([| for e in workspace.Effects -> EffectViewModel(e, messenger) |])

    member _.Effects = lazyEffect.Value
